File: packages/swift-client/Sources/PyricFirestore/Public/DocumentChange.swift
import Foundation

public enum DocumentChangeType: Int, Sendable, Equatable {
    case added = 0
    case modified = 1
    case removed = 2
}

public final class DocumentChange: Sendable, Equatable {
    public let type: DocumentChangeType
    public let document: QueryDocumentSnapshot
    public let oldIndex: Int
    public let newIndex: Int

    public static let indexNotFound: Int = -1

    public init(
        type: DocumentChangeType,
        document: QueryDocumentSnapshot,
        oldIndex: Int,
        newIndex: Int
    ) {
        self.type = type
        self.document = document
        self.oldIndex = oldIndex
        self.newIndex = newIndex
    }

    public static func == (lhs: DocumentChange, rhs: DocumentChange) -> Bool {
        lhs.type == rhs.type &&
        lhs.oldIndex == rhs.oldIndex &&
        lhs.newIndex == rhs.newIndex &&
        lhs.document.documentID == rhs.document.documentID
    }
}

File: packages/kt-client/src/main/kotlin/com/google/firebase/firestore/DocumentChange.kt
package com.google.firebase.firestore

class DocumentChange internal constructor(
    val type: Type,
    val document: QueryDocumentSnapshot,
    val oldIndex: Int,
    val newIndex: Int
) {
    enum class Type {
        ADDED,
        MODIFIED,
        REMOVED
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is DocumentChange) return false
        return type == other.type && document == other.document && oldIndex == other.oldIndex && newIndex == other.newIndex
    }

    override fun hashCode(): Int {
        var result = type.hashCode()
        result = 31 * result + document.hashCode()
        result = 31 * result + oldIndex
        result = 31 * result + newIndex
        return result
    }

    override fun toString(): String =
        "DocumentChange(type=$type, document=${document.id}, oldIndex=$oldIndex, newIndex=$newIndex)"
}
